p2p/chain.py

```diff
@@ -137,17 +146,66 @@ def register_peer(self, peer: BasePeer) -> None:
                 continue
 
             # TODO: Process headers for consistency.
-            # TODO: Queue body/receipt downloads.
             for header in headers:
                 await self.chaindb.coro_persist_header(header)
                 start_at = header.block_number
 
+            # TODO: Queue receipt downloads.
+            self._body_requests.put_nowait(headers)
+
             self.logger.debug("Asking %s for header batch starting at %d", peer, start_at)
             # TODO: Instead of requesting sequential batches from a single peer, request a header
             # skeleton and make concurrent requests, using as many peers as possible, to fill the
             # skeleton.
             peer.sub_proto.send_get_block_headers(start_at, eth.MAX_HEADERS_FETCH, reverse=False)
 
+    async def body_downloader(self) -> None:
+        while True:
+            try:
+                headers = await wait_with_token(
+                    self._body_requests.get(),
+                    token=self.cancel_token,
+                    timeout=self._reply_timeout)
+            except TimeoutError:
+                # We use a timeout above just to make sure we periodically retry timedout body
+                # requests even when there's no new bodies coming through.
+                headers = []
+            except OperationCancelled:
+                return
+
+            if headers:
+                headers_with_non_empty_bodies = [
+                    header for header in headers
+                    if (header.transaction_root != self.chaindb.empty_root_hash or
+                        header.uncles_hash != EMPTY_UNCLE_HASH)
+                ]
+                self.request_bodies(headers_with_non_empty_bodies)
+
+            self.retry_timedout_bodies()
+
+    def request_bodies(self, headers: List[BlockHeader]) -> None:
+        while headers:
```

Comment:
This body can be cleaned up a bit by using [`cytoolz.partition_all`](http://toolz.readthedocs.io/en/latest/api.html#toolz.itertoolz.partition_all)

```python
for batch in partition_all(headers, eth.MAX_BODIES_FETCH):
    ...
```

Nice side effect of removing the mutation of the `headers` variable.

Reply:
Whenever I do something like this I remember to look at itertools to see if there's something I could use, but I never remember to look at cytoolz as well. Thanks for the reminder!

Reply:
FYI, this is the documentation page you should bookmark: http://toolz.readthedocs.io/en/latest/api.html